Core/include/Acts/Utilities/Any.hpp

```diff
@@ -0,0 +1,495 @@
+#endif
+
+#if defined(_ACTS_ANY_ENABLE_DEBUG)
+#define _ACTS_ANY_DEBUG(x) std::cout << x << std::endl;
+#else
+#define _ACTS_ANY_DEBUG(x)
+#endif
+
+#if defined(_ACTS_ANY_ENABLE_VERBOSE)
+#define _ACTS_ANY_VERBOSE(x) std::cout << x << std::endl;
+#define _ACTS_ANY_VERBOSE_BUFFER(s, b) \
+  do {                                 \
+    std::cout << "" << s << ": 0x";    \
+    for (char c : b) {                 \
+      std::cout << std::hex << (int)c; \
+    }                                  \
+    std::cout << std::endl;            \
+  } while (0)
+#else
+#define _ACTS_ANY_VERBOSE(x)
+#define _ACTS_ANY_VERBOSE_BUFFER(s, b)
+#endif
+
+namespace Acts {
+
+#if defined(_ACTS_ANY_ENABLE_TRACK_ALLOCATIONS)
+static std::mutex _s_any_mutex;
+static std::set<std::pair<std::type_index, void*>> _s_any_allocations;
+
+#define _ACTS_ANY_TRACK_ALLOCATION(T, heap)                                  \
+  do {                                                                       \
+    std::lock_guard guard{_s_any_mutex};                                     \
+    _s_any_allocations.emplace(std::type_index(typeid(T)), heap);            \
+    _ACTS_ANY_DEBUG("Allocate type: " << typeid(T).name() << " at " << heap) \
+  } while (0)
+
+struct _AnyAllocationReporter {
+  ~_AnyAllocationReporter() {
+    std::lock_guard guard{_s_any_mutex};
+
+    if (!_s_any_allocations.empty()) {
+      std::cout << "Not all allocations have been released" << std::endl;
+      for (const auto& [idx, addr] : _s_any_allocations) {
+        std::cout << "- " << idx.name() << ": " << addr << std::endl;
+      }
+      std::abort();
+    }
+  }
+};
+static _AnyAllocationReporter s_reporter;
+#else
+#define _ACTS_ANY_TRACK_ALLOCATION(T, heap)
+#endif
+
+class AnyBaseAll {};
+
+/// Small opaque cache type which uses small buffer optimization
+template <size_t SIZE>
+class AnyBase : public AnyBaseAll {
+  static_assert(sizeof(void*) <= SIZE, "Size is too small for a pointer");
+
+ public:
+  template <typename T, typename... Args>
+  explicit AnyBase(std::in_place_type_t<T> /*unused*/, Args&&... args) {
+    using U = std::decay_t<T>;
+    static_assert(
+        std::is_move_assignable_v<U> && std::is_move_constructible_v<U>,
+        "Type needs to be move assignable and move constructible");
+    static_assert(
+        std::is_copy_assignable_v<U> && std::is_copy_constructible_v<U>,
+        "Type needs to be copy assignable and copy constructible");
+
+    m_handler = makeHandler<U>();
+    if constexpr (sizeof(U) <= SIZE) {
+      // construct into local buffer
+      /*U* ptr =*/new (m_data.data()) U(std::forward<Args>(args)...);
+      _ACTS_ANY_VERBOSE(
+          "Construct local (this=" << this << ") at: " << (void*)m_data.data());
+    } else {
+      // too large, heap allocate
+      U* heap = new U(std::forward<Args>(args)...);
+      _ACTS_ANY_TRACK_ALLOCATION(T, heap);
+      setDataPtr(heap);
+    }
+  }
+
+  AnyBase() {
+    m_data.fill(std::byte{0});
```

Comment:
Why?

Reply:
I guess so it's a nullptr by default. Might not be necessary

Reply:
Well, your null state is really indicated by `m_handler` being `nullptr`, and it has to be because with SBO, you need to check `m_handler` in order to correctly interpret the value of `m_data` (it could be a perfectly valid array of zero bytes after all). So I would argue that this is unnecessary.